server/base/src/main/java/org/apache/accumulo/server/util/MasterMetadataUtil.java

```diff
@@ -247,61 +243,58 @@ public static void updateTabletDataFile(ServerContext context, KeyExtent extent,
     }
-    Mutation m = getUpdateForTabletDataFile(extent, path, mergeFile, dfv, time, filesInUseByScans,
-        address, zooLock, unusedWalLogs, lastLocation, flushId);
-    MetadataTableUtil.update(context, zooLock, m, extent);
+
   }
 
   /**
    * Update the data file for the root tablet
    */
   private static void updateRootTabletDataFile(ServerContext context, Set<String> unusedWalLogs) {
-    TabletMutator tablet = context.getAmple().mutateTablet(RootTable.EXTENT);
-    unusedWalLogs.forEach(tablet::deleteWal);
-    tablet.mutate();
+    if (unusedWalLogs != null) {
+      TabletMutator tablet = context.getAmple().mutateTablet(RootTable.EXTENT);
+      unusedWalLogs.forEach(tablet::deleteWal);
+      tablet.mutate();
+    }
   }
 
   /**
    * Create an update that updates a tablet
    *
-   * @return A Mutation to update a tablet from the given information
    */
-  private static Mutation getUpdateForTabletDataFile(KeyExtent extent, FileRef path,
+  private static void updateForTabletDataFile(ServerContext context, KeyExtent extent, FileRef path,
       FileRef mergeFile, DataFileValue dfv, String time, Set<FileRef> filesInUseByScans,
       String address, ZooLock zooLock, Set<String> unusedWalLogs, TServerInstance lastLocation,
       long flushId) {
-    Mutation m = new Mutation(extent.getMetadataEntry());
+
+    TabletMutator tablet = context.getAmple().mutateTablet(extent);
 
     if (dfv.getNumEntries() > 0) {
-      m.put(DataFileColumnFamily.NAME, path.meta(), new Value(dfv.encode()));
-      TabletsSection.ServerColumnFamily.TIME_COLUMN.put(m, new Value(time.getBytes(UTF_8)));
-      // stuff in this location
+      tablet.putFile(path, dfv);
+      tablet.putTime(time);
+
       TServerInstance self = getTServerInstance(address, zooLock);
-      self.putLastLocation(m);
-      // erase the old location
-      if (lastLocation != null && !lastLocation.equals(self))
-        lastLocation.clearLastLocation(m);
-    }
-    if (unusedWalLogs != null) {
-      for (String entry : unusedWalLogs) {
-        m.putDelete(LogColumnFamily.NAME, new Text(entry));
+      tablet.putLocation(self, LocationType.LAST);
+
+      // remove the old location
+      if (lastLocation != null && !lastLocation.equals(self)) {
+        tablet.deleteLocation(lastLocation, LocationType.LAST);
       }
     }
+    tablet.putFlushId(flushId);
 
-    for (FileRef scanFile : filesInUseByScans)
-      m.put(ScanFileColumnFamily.NAME, scanFile.meta(), new Value(new byte[0]));
-
-    if (mergeFile != null)
-      m.putDelete(DataFileColumnFamily.NAME, mergeFile.meta());
-
-    TabletsSection.ServerColumnFamily.FLUSH_COLUMN.put(m,
-        new Value(Long.toString(flushId).getBytes(UTF_8)));
+    if (mergeFile != null) {
+      tablet.deleteFile(mergeFile);
+    }
+    if (unusedWalLogs != null) {
+      unusedWalLogs.forEach(tablet::deleteWal);
+    }
+    filesInUseByScans.forEach(tablet::putScan);
 
-    return m;
+    tablet.mutate();
```

Comment:
This method should call `tablet.putZooLock(zooLock);`  to preserve existing behavior.

Reply:
yes.  I missed that it was being done in the MetadataTableUtil.update()